Stop command

michele-brambilla opened the pull request:
Stop the writing of a (specific) file but keep the filewriter active so that is can continue writing other files or accept new request. A unique jobid as part of the command to start writing a file is expected; if not present uses id field in FileWriterTask.

matthew-d-jones commented on src/tests/msg-conf-new-01.json:
```diff
@@ -1,6 +1,7 @@
 {
 	"cmd": "FileWriter_new",
 	"broker": "localhost:9092",
+        "jobid": "0000000000000001", 
```
Is this then used in a system test?
It would be straightforward and beneficial to also implement unit tests for the parsing of command messages.

michele-brambilla replied:
Yes, it is used in ```roundtrip.cxx```. I added some tests for the new functions ```find_*``` used during parsing, others are tested in ```T_CommandHandler```. I'll see if something else can be extracted and tested

michele-brambilla commented on src/CommandHandler.cxx:
```diff
@@ -54,6 +54,23 @@ void CommandHandler::handle_new(rapidjson::Document const &d) {
+    }
+  }
+  // how to handle missing jobid?
+  fwt->jobid_init(jobid);
```
So far it is a request that Nikhil made to allow Nikos to stop writing one file without affecting the filewriter itself (that runs as a service). Should it be a hard requirement? Probably. @dominikwerder , @nikhilbiyani what do you think about?

michele-brambilla replied:
So far it is a request that Nikhil made to allow Nikos to stop writing one file without affecting the filewriter itself (that runs as a service). Should it be a hard requirement? Probably. @dominikwerder , @nikhilbiyani what do you think about?

matthew-d-jones replied:
@michele-brambilla I just mean to suggest that it should give up on parsing the rest of this command message and go back to listening for another one.

michele-brambilla replied:
:thumbsup:

nikhilbiyani replied:
I think that job_id can be made compulsory. Otherwise it gets hard for to track the jobs.

michele-brambilla replied:
:white_check_mark:

nikhilbiyani commented on README.md:
````diff
@@ -112,6 +113,15 @@ Command to exit the file writer:
+	"jobid": "job-unique-identifier"
+}
+```
+
````
Should we provide an optional `stop_time` here?

michele-brambilla replied:
:white_check_mark:

michele-brambilla commented on README.md:
````diff
@@ -112,6 +113,15 @@ Command to exit the file writer:
+```json
+{
+	"cmd": "FileWriter_stop",
+	"jobid": "job-unique-identifier"
````
:white_check_mark:

michele-brambilla commented:
Changed according to reviews:
- [x] if command doesn't contain ``job_id`` is ignored 
- [x] ``handle_new`` calls ``find_XXX`` functions
- [x] added unit tests for the parsing of command messages

@matthew-d-jones: this PR fixes the bug in #38, nevertheless the separation between unit test and integration test is still an open issue. Should I mark this PR as closing #38 anyway?
@dominikwerder : can you let me know if you like how ``handle_new`` uses ``find_XXX`` ?

matthew-d-jones commented:
@michele-brambilla You could mark as closing #38 and create a separate issue for separating unit and integration tests.
Should I take another look through this PR or are you making further changes?

michele-brambilla commented:
No, I'm not going to make any change, please have a look

matthew-d-jones commented on src/tests/command_handler.cxx:
```diff
@@ -0,0 +1,100 @@
+  }
+};
+
+TEST_F(CommandHandler_Test, filename) {
```
Running unit tests can catch cases where the output is not what the programmer intended. However another cause of bugs is actually that the programmer was mistaken about what the output should be. An important role of unit tests is therefore to document the intention of the programmer so that other people can spot such mistakes. For this reason I think it is a really good idea to have readable test names which make it clear what is being tested. As an example, this test could be called `test_found_filename_matches_that_in_json`. Consider renaming this and the other tests.

michele-brambilla commented on src/DemuxTopic.h:
```diff
@@ -40,12 +41,12 @@ class DemuxTopic : public TimeDifferenceFromMessage, public MessageProcessor {
   rapidjson::Document
   to_json(rapidjson::MemoryPoolAllocator<> *_a = nullptr) const;
-  uint64_t &stop_time();
+  ESSTimeStamp &stop_time();
```
Yes, ``StreamMaster`` uses it set the time at which the writing has to stop. It s useful _e.g._ for scanning or measurements whose duration can/has to be set in advance.

michele-brambilla commented on src/tests/command_handler.cxx:
```diff
@@ -0,0 +1,100 @@
+std::string find_filename(rapidjson::Document const &);
+std::string find_job_id(rapidjson::Document const &);
+std::string find_broker(rapidjson::Document const &);
+std::vector<std::unique_ptr<Source>> find_source(rapidjson::Document const &,
```
I tried to remove the line, but the building fails

matthew-d-jones replied:
Ah I see, because it isn't declared in `CommandHandler.h`.

michele-brambilla commented on src/CommandHandler.cxx:
```diff
@@ -2,8 +2,79 @@
+      uri::URI u(s);
+      return u.host_port;
+    } else {
+      // legacy semantics
```
If you mean get rid of the ``//legacy semantics`` part I'm not sure. The choice here depends on how we want to describe the broker in the command file. Will it be part of the ``topic`` field or will it be a field on its own? What if we keep it for the moment (maybe we can open an issue) and we wait until we discuss with the rest of the collaboration?

matthew-d-jones replied:
It is not clear to me what your "// legacy semantics" comment means, is it because you have changed the json format?
If so, my point is that we really should not be afraid of losing support for a format which no one is using yet anyway.

matthew-d-jones replied:
I see, if the decision is to support both then it is fine, but the comment is misleading.
"legacy" implied keeping something just for backwards compatibility, this is what concerned me.

michele-brambilla commented:
I made the correction you request: 
[x] removed dead code
[x] fixed pass-by-reference and return values
[x] minor corrections after merge
If you want to give a further look I don't have any other changes to push
